edge/pkg/devicetwin/dmiclient/client.go

```diff
@@ -0,0 +1,310 @@
+limitations under the License.
+*/
+
+package dmiclient
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"k8s.io/klog/v2"
+
+	deviceconst "github.com/kubeedge/kubeedge/cloud/pkg/devicecontroller/constants"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/pkg/apis/devices/v1alpha2"
+	dmiapi "github.com/kubeedge/kubeedge/pkg/apis/dmi/v1"
+)
+
+type DMIClient struct {
+	protocol   string
+	socket     string
+	Client     dmiapi.DeviceMapperServiceClient
+	Ctx        context.Context
+	Conn       *grpc.ClientConn
+	CancelFunc context.CancelFunc
+}
+
+type DMIClients struct {
+	mutex   sync.Mutex
+	clients map[string]*DMIClient
+}
+
+var DMIClientsImp *DMIClients
+
+func init() {
+	DMIClientsImp = &DMIClients{
+		mutex:   sync.Mutex{},
+		clients: make(map[string]*DMIClient),
+	}
+}
+
+func (dc *DMIClient) connect() error {
+	dialer := func(addr string, t time.Duration) (net.Conn, error) {
+		return net.Dial(deviceconst.UnixNetworkType, addr)
+	}
+
+	conn, err := grpc.Dial(dc.socket, grpc.WithInsecure(), grpc.WithDialer(dialer))
+	if err != nil {
+		klog.Errorf("did not connect: %v\n", err)
+		return err
+	}
+
+	c := dmiapi.NewMapperClient(conn)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+
+	dc.Client = c
+	dc.Ctx = ctx
+	dc.Conn = conn
+	dc.CancelFunc = cancel
+
+	return nil
+}
+
+func (dc *DMIClient) close() {
+	dc.Conn.Close()
+	dc.CancelFunc()
+}
+
+func createDeviceRequest(device *v1alpha2.Device) (*dmiapi.CreateDeviceRequest, error) {
+	d, err := dtcommon.ConvertDevice(device)
+	if err != nil {
+		return nil, err
+	}
+
+	return &dmiapi.CreateDeviceRequest{
+		Device: d,
+	}, nil
+}
+
+func removeDeviceRequest(deviceName string) (*dmiapi.RemoveDeviceRequest, error) {
+	return &dmiapi.RemoveDeviceRequest{
+		DeviceName: deviceName,
+	}, nil
+}
+
+func updateDeviceRequest(device *v1alpha2.Device) (*dmiapi.UpdateDeviceRequest, error) {
+	d, err := dtcommon.ConvertDevice(device)
+	if err != nil {
+		return nil, err
+	}
+
+	return &dmiapi.UpdateDeviceRequest{
+		Device: d,
+	}, nil
+}
+
+func createDeviceModelRequest(model *v1alpha2.DeviceModel) (*dmiapi.CreateDeviceModelRequest, error) {
+	m, err := dtcommon.ConvertDeviceModel(model)
+	if err != nil {
+		return nil, err
+	}
+
+	return &dmiapi.CreateDeviceModelRequest{
+		Model: m,
+	}, nil
+}
+
+func updateDeviceModelRequest(model *v1alpha2.DeviceModel) (*dmiapi.UpdateDeviceModelRequest, error) {
+	m, err := dtcommon.ConvertDeviceModel(model)
+	if err != nil {
+		return nil, err
+	}
+
+	return &dmiapi.UpdateDeviceModelRequest{
+		Model: m,
+	}, nil
+}
+
+func removeDeviceModelRequest(deviceModelName string) (*dmiapi.RemoveDeviceModelRequest, error) {
+	return &dmiapi.RemoveDeviceModelRequest{
+		ModelName: deviceModelName,
+	}, nil
+}
+
+func (dcs *DMIClients) getDMIClientByProtocol(protocol string) (*DMIClient, error) {
+	dc, ok := dcs.clients[protocol]
```

Comment:
dcs.mutex.Lock()

Reply:
and one client for one protocol ? shouldn't it be one client for one device?(how to deal devices with different address and with the same protocol, i think on the scene, the only one client is not right?)

Reply:
> dcs.mutex.Lock()

done

Reply:
> and one client for one protocol ? shouldn't it be one client for one device?(how to deal devices with different address and with the same protocol, i think on the scene, the only one client is not right?)

one client for one protocol and one mapper. if you want to create two devices with the same protocol, just use the same client to call `CreateDevice` twice.